opnode/l1/source.go

```diff
@@ -74,8 +76,19 @@ func (s Source) Close() {
 func (s Source) FetchL1Info(ctx context.Context, id eth.BlockID) (derive.L1Info, error) {
 	return s.client.BlockByHash(ctx, id.Hash)
 }
-func (s Source) FetchReceipts(ctx context.Context, id eth.BlockID) ([]*types.Receipt, error) {
+
+func (s Source) FetchReceipts(ctx context.Context, id eth.BlockID, receiptHash common.Hash) ([]*types.Receipt, error) {
 	_, receipts, err := s.Fetch(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	// Sanity-check: external L1-RPC sources are notorious for not returning all receipts,
+	// or returning them out-of-order. Verify the receipts against the expected receipt-hash.
+	hasher := trie.NewStackTrie(nil)
+	computed := types.DeriveSha(types.Receipts(receipts), hasher)
+	if receiptHash != computed {
```

Comment:
I like this. Do we need to implement a retry loop?
Also, does this mean that we may need to do a cache purge on this if we don't compute the correct result?

Reply:
The cache could wrap this, so it never caches when the error hits. Will look at that in the upcoming downloader PR. I don't think we should implement a retry-loop, the error will just appear like any other receipt retrieval issue, and it will retry later. Good point about cache purging though :+1:

Reply:
Sounds good to rely on higher level retry.